### PR TITLE
Split up PORTER_HOME 

### LINKED ISSUE #1328: Split up PORTER_HOME
Currently PORTER_HOME (defaults to ~/.porter) contains all kinds of files: porter binaries, mixins, plugins, cache, bundle installation data, etc. This doesn't follow standards for separating installation data from user data and is a problem for us packaging up [Porter for Homebrew](https://github.com/getporter/porter/discussions/1310) (and will probably be a problem for Chocolatey as well).

Let's split up the where we store data based on its type:

## System
The installation of Porter to the system. Depending on how it is installed: homebrew, chocolately, manual download, script; it will be a different location. For example homebrew's system directory for Porter would be `/usr/local/Cellar/porter/<version>/`. This would contain the following files:

* porter

## User
These are files specific to the user and can vary when there are multiple users on a machine. The location is determined using a hierarchy of checks:

1. PORTER_HOME
1. [XDG_DATA_HOME](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html)
1. HOME/.porter or USERPROFILE/.porter

If none of these environment variables are defined, it is an error condition. We are not going to guess where your home directory is and require that you set PORTER_HOME when nothing else is available.

* runtimes/porter-linux-amd64 (previously porter-runtime*)
* mixins
* plugins
* cache (When XDG is defined, we should use XDG_CACHE_HOME)
* config.* (When XDG is defined, we should use XDG_CONFIG_HOME)

\* This needs to move because eventually we are going to support multiple invocation image runtime formats, and looking up the location of the runtime binary from the user directory is where it's going to end up eventually anyway. We may need to do this separately but I wanted to call out that it may be more work to keep it in system.

## Data
These are data files for working with installations. They are logically separate because they may not even be stored on the filesystem. This data is processed through the storage plugin to allow for sharing with other people via a shared location in the cloud. When they are stored on the filesystem, they go in PORTER_HOME.

* installations
* claims
* results
* outputs
* credentials
* parameters

# Notes

The changes necessary to support this are mostly in place. For example, when Porter needs to find a mixin, it looks for the porter home directory first, then /mixins/MIXIN. Same for plugins, and the cache. What needs to change is:

* makefile install target
* Our installation script
* [pkg/config.GetHomeDir](https://github.com/getporter/porter/blob/2fcaeb4b1490adaf9e33a2af1dedfa7c93413251/pkg/config/config.go#L86) which is used to find the user directory (e.g. PORTER_HOME)
* [pkg/cache.GetCacheDir](https://github.com/getporter/porter/blob/2fcaeb4b1490adaf9e33a2af1dedfa7c93413251/pkg/cache/cache.go#L162)
* [Reading config file](https://github.com/getporter/porter/blob/2fcaeb4b1490adaf9e33a2af1dedfa7c93413251/pkg/config/datastore/loader.go#L46)
* Resolution of porter binary, [pkg/config.GetPorterPath](https://github.com/getporter/porter/blob/2fcaeb4b1490adaf9e33a2af1dedfa7c93413251/pkg/config/config.go#L129)
* Resolution of porter-runtime binary, e.g. [pkg/config.GetPorterRuntimePath](https://github.com/getporter/porter/blob/2fcaeb4b1490adaf9e33a2af1dedfa7c93413251/pkg/config/config.go#L139)

### PR DESCRIPTION
**Pull request by sestegra**
# What does this change
Split up the where we store porter executable and data based on its type.

# What issue does it fix
Closes #1328 

# Tasks
- [X] Installation scripts
- [x] Makefile install target
- [x] Compute home dir
- [X] Resolution of porter binary
- [X] Resolution of porter-runtime binary

# Notes for the reviewer
- For installation scripts, I suggest to keep porter executable in `$HOME/.porter` directory for now, because system installation would need sudo permission.
- PORTER_HOME should be exposed in Docker image.
- `runtimes` or `runtime` without `s`?

# Checklist
- [x] Unit Tests
- [x] Documentation

### CODE REVIEW

**Comment by sestegra** [comment]
@carolynvs I'm struggling with integration-test and validate_example_bundles that don't pass.

**Review by carolynvs — COMMENTED** [review]
This is looking real good and I think you are super close! 💯 

One aspect that I missed when outlining what should change is that the porter binary is built and staged the same as the mixin binaries. For example, porter uses the mixin.mk to build itself and that makefile is still placing the output binaries for the runtime next to the client binary.

https://github.com/getporter/porter/blob/a8bb979abb99f7368972dad905f04149c6fc4e59/mixin.mk#L35-L37

Since mixins too will need to support multiple runtime OS/ARCH requirements soon, let's make the same change to their directory structure as well and that should resolve the problem.

So after a build we should see (below), and now the changes you have made in the code will be supported by what the build is outputting to bin.

```
bin/
  mixins/
    exec/
      exec
      runtimes/
        exec-runtime
  runtimes/
    porter-runtime
  porter
```

I recommend looking for "-runtime" in the code and fixing up the path so that we are looking in the runtimes/ directory. I think there are 1-2 more places besides this, maybe when we build the dockerfile that would also need to be fixed. For example, here is where we resolve the path for a mixin's binary

https://github.com/getporter/porter/blob/a8bb979abb99f7368972dad905f04149c6fc4e59/pkg/pkgmgmt/client/runner.go#L102-L108

File: Makefile [hunk 1]
```diff
@@ -208,9 +208,9 @@ endif
 install: install-porter install-mixins
 
 install-porter:
-	mkdir -p $(HOME)/.porter
-	cp bin/porter* $(HOME)/.porter/
-	ln -f -s $(HOME)/.porter/porter /usr/local/bin/porter
+	mkdir -p $(HOME)/.porter/runtime
```
[the remark this answers is not in the document]
Comment (sestegra) [reply Makefile]: Done

File: pkg/config/config.go [hunk 2]
```diff
@@ -127,22 +117,34 @@ func (c *Config) SetHomeDir(home string) {
...
 func (c *Config) GetPorterRuntimePath() (string, error) {
-	path, err := c.GetPorterPath()
+	home, err := c.GetHomeDir()
 	if err != nil {
 		return "", nil
 	}
 
-	return path + "-runtime", nil
+	runtimePath := filepath.Join(home, "runtime")
+	runtimeBinaryPath := filepath.Join(runtimePath, "porter-runtime")
```
[the remark this answers is not in the document]
Comment (sestegra) [reply pkg/config/config.go:146]: Done

**Comment by sestegra** [comment]
@carolynvs 
It seems that the missing issue on test is a chicken or the egg about `TestArchive` that needs an upload of a package that can't be at the right format.

**Comment by sestegra** [comment]
Issue linked to #697

**Comment by carolynvs** [comment]
I'll look into how we can get the tests passing and get back to you.

**Comment by carolynvs** [comment]
I am fixing some problems I found in Porter that I think are contributing to the test errors. I have #1345 open and am working to clear those obstacles for this PR. Just wanted you to know that I'm working on it. Once that PR is merged, yours should need just a few tweaks to pass the tests.

**Comment by carolynvs** [comment]
OK! Once #1345 is merged tomorrow, you can rebase/merge the main branch to get the changes. I have also made [3 small changes](https://github.com/carolynvs/porter/commit/e7631975dd47d192e19fd24a96d4afacab0223dc) that got your pull request to pass the integration tests. Copy those changes into your branch and we should be all set.

**Comment by sestegra** [comment]
Merges done. Thank you for your support.

File: CONTRIBUTORS.md [hunk 3]
```diff
@@ -48,4 +48,5 @@ and we will add you. **All** contributors belong here. 💯
 * [Michael William Boldt](https://github.com/mboldt)
 * [Mohamed Chorfa](https://github.com/MChorfa)
 * [Don Stewart](https://github.com/donmstewart)
-* [Nicholas Carrigan](https://github.com/nhcarrigan)
\ No newline at end of file
+* [Nicholas Carrigan](https://github.com/nhcarrigan)
+* [Stéphane Este-Gracias](https://github.com/sestegra)
```
Comment (carolynvs) [inline CONTRIBUTORS.md:52]: 🎉

**Review by carolynvs — APPROVED** [review]
This is just great! Thanks so much for both the designs and getting it implemented. 🙇‍♀️

**Pull request merged**